### PR TITLE
feature/schematics

### PR DESCRIPTION
**Pull request by BeastCode**
Work in progress, first PR for schematics.

# Install Schematics

npm i -g @angular-devkit/{schematics, schematics-cli, core } @shematics/schematics

# Debug and Try Ngxs Schematics

schematics .:state --name=animals --dry-run=false
schematics .:action --name=FeedAnimals --state=Zoo --dry-run=false
schematics .:select --name=Zoo --dry-run=false

### CODE REVIEW

**Comment by leon** [comment]
Hi!
Could you please refactor and put the schematics in it's own folder under packages.
Make a copy of one of the plugins and change the package.json and ng-packagr.json files to have `@ngxs/schematics`.

and to get the tooling to build the schematics, add `schematics` to the `package.json` packages key.

Let me know if you need any more guidance :)

**Comment by amcdnl** [comment]
@leon - I asked him to put it under the store folder. Reason is there is a `ng add` functionality that is not documented that allows you to add a new package and hook up all its deps but the schematics need to live in that package to do so.

File: packages/store/src/schematics/src/state/index.ts [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+      template({
+        ...stringUtils,
+        name: options.name,
+      }),
+    ])),
+  ]);
+}
+
+/*
+TODO
```
Comment (amcdnl) [inline packages/store/src/schematics/src/state/index.ts]: See this - https://github.com/angular/material2/blob/0a501586a5a72feb924dd3b2b3bae890454bd1cc/src/lib/schematics/utils/ast.ts#L26-L53

File: packages/store/src/schematics/src/collection.json [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+  "$schema": "./node_modules/@angular-devkit/schematics/collection-schema.json",
+
+  // Schematics are listed as a map of schematicName => schematicDescription.
+  // Each description contains a description field which is required, a factory reference,
+  // an extends field and a schema reference.
+  // The extends field points to another schematic (either in the same collection or a
+  // separate collection using the format collectionName:schematicName).
+  // The factory is required, except when using the extends field. Then the factory can
+  // overwrite the extended schematic factory.
+  "schematics": {
```
Comment (amcdnl) [inline packages/store/src/schematics/src/collection.json]: We probably want a `ng add` schematic. See here for more details - https://github.com/angular/material2/blob/0a501586a5a72feb924dd3b2b3bae890454bd1cc/src/lib/schematics/collection.json#L6

File: packages/store/src/schematics/README.md [hunk 3]
```diff
@@ -0,0 +1,40 @@
+# Work in Progress
```
Comment (amcdnl) [inline packages/store/src/schematics/README.md]: This should be in the docs/advanced/schematics.md

File: packages/store/src/schematics/src/select/files/__name@dasherize__.select.ts [hunk 4]
```diff
@@ -0,0 +1,10 @@
+import { Select } from '@ngxs/store';
+ 
+import { <%= classify(name) %>State } from './<%= dasherize(name) %>.state';
+import { animal } from './animal.actions';
+ 
+@Component({ ... })
```
Comment (amcdnl) [inline packages/store/src/schematics/src/select/files/__name@dasherize__.select.ts]: I wonder if we really need this schematic.

**Comment by amcdnl** [comment]
I'm thinking there might be 2 schematics -

- `ng add @ngxs/store` - Install deps and add to root module
- `ng  g state zoo`

**Comment by deebloo** [comment]
I think we are going to need to ignore these ts files in the tsconfig or things will continue to break

File: packages/store/src/schematics/package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+  "schematics": "./src/collection.json",
+  "dependencies": {
+    "@angular-devkit/core": "^0.5.4",
+    "@angular-devkit/schematics": "^0.5.4",
+    "@types/jasmine": "^2.6.0",
+    "@types/node": "^8.0.31",
+    "jasmine": "^2.8.0",
+    "typescript": "^2.5.2"
+  }
+}
```
Comment (deebloo) [inline packages/store/src/schematics/package.json]: do we need a separate package.json for this? or can it be in the root for this module

File: packages/store/src/schematics/src/collection.json (hunk 2 above)
Comment (deebloo) [inline packages/store/src/schematics/src/collection.json]: won't it break if we have comments in the json file?

File: packages/store/src/schematics/tsconfig.json [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      "node"
+    ]
+  },
+  "include": [
+    "src/**/*"
+  ],
+  "exclude": [
+    "src/*/files/**/*"
+  ]
+}
```
Comment (deebloo) [inline packages/store/src/schematics/tsconfig.json]: why is there a separate tsconfig just for schematics?

**Comment by deebloo** [comment]
@BeastCode added some comments/questions

File: packages/store/src/schematics/.gitignore [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+.vscode/
+
+# Misc
+node_modules/
+npm-debug.log*
+yarn-error.log*
+
+# Mac OSX Finder files.
+**/.DS_Store
+.DS_Store
```
Comment (deebloo) [inline packages/store/src/schematics/.gitignore]: I don't think this file needs to be here

File: packages/store/src/schematics/src/collection.json (hunk 2 above)
Comment (amcdnl) [reply packages/store/src/schematics/src/collection.json]: JSON5 supports it

File: packages/store/src/schematics/package.json (hunk 5 above)
Comment (amcdnl) [reply packages/store/src/schematics/package.json]: Yes.

**Comment by amcdnl** [comment]
Gonna close this for now.

**Comment by mackelito** [comment]
I know this is closed but I'm just wondering if @ngxs/store support schematics?

**Pull request closed without merging**